### ISSUE TITLE
rosbash need to be updated for newer version of fish shell

### ISSUE THREAD

**Issue by NextAlone** [ros/ros#286]

**Comment by hidmic** [comment]
@NextAlone could you be bit more specific? are you willing to contribute?

**Comment by NextAlone** [comment]
> @NextAlone could you be bit more specific? are you willing to contribute?

Such as `rostopic pub /vel_cmd geometry/Twist`
It was parsed wrongly with "fish shell string utility with new versions"

Later maybe have some free time to contribute

**Comment by NextAlone** [comment]
> @NextAlone could you be bit more specific? are you willing to contribute?

Such as `rostopic pub /vel_cmd geometry/Twist`
It was parsed wrongly with "fish shell string utility with new versions"

Later maybe have some free time to contribute

**Comment by NextAlone** [comment]
rostopic pub /rosout rosgraph_msgs/Log  (press tab)

(complete as follows)
![image](https://user-images.githubusercontent.com/12210746/145072583-9a420b90-8dd9-4541-b679-51ac48045d6a.png)

(it should be what follows in zsh )
![image](https://user-images.githubusercontent.com/12210746/145072760-2dc984b6-f79e-43a8-bdb4-fd9f3ed21b16.png)

the content complete as string and can be parsed by fish to split

**Comment by hidmic** [comment]
> Later maybe have some free time to contribute

That would be most appreciated @NextAlone.

**Comment by NextAlone** [comment]
For now only find that `rostopic pub` has error complete. Only to manage to be like this, which is completed.
```
rostopic pub /rosout rosgraph_msgs/Log \{\"header\":\{\"seq\":0,\"stamp\":\{\"secs\":0,\"nsecs\":0\},\"frame_id\":\"\"\},\"level\":0,\"name\":\"\",\"msg\":\"\",\"file\":\"\",\"function\":\"\",\"line\":0,\"topics\":\[\"\"\]\}
```
compared to what like in zsh
```
rostopic pub /rosout rosgraph_msgs/Log "header:                                                                                                                                        :(
  seq: 0
  stamp: {secs: 0, nsecs: 0}
  frame_id: ''
level: 0
name: ''
msg: ''
file: ''
function: ''
line: 0
topics: ['']"
```

Code diff is as follow. Add yq and jq as dependencies for YAML parse and json generate. 
https://github.com/ros/ros/blob/93d8da32091b8b43702eab5d3202f4511dfeb7dc/tools/rosbash/rosfish#L949
```
-        set -l opts (rosmsg-proto msg 2> /dev/null -s $command[-2])
+        set -l opts (rosmsg-proto msg 2> /dev/null -s $command[-2] | string replace "\"" "" | yq | jq -c)
```

**Comment by DragonflyRobotics** [comment]
The output looks fine to me in bash. I am also running ROS Noetic. I checked on both zsh and bash. Look at the following screenshots:
![image](https://user-images.githubusercontent.com/85193239/158045185-987838dc-df9c-47af-821f-5a4782151c99.png)

![image](https://user-images.githubusercontent.com/85193239/158045203-d93180c2-744c-4e17-af9b-d90b9fa8df72.png)

I am not sure if this has been patched already.

**Comment by NextAlone** [comment]
> The output looks fine to me in bash. I am also running ROS Noetic. I checked on both zsh and bash. Look at the following screenshots:
> ![image](https://user-images.githubusercontent.com/85193239/158045185-987838dc-df9c-47af-821f-5a4782151c99.png)
> 
> ![image](https://user-images.githubusercontent.com/85193239/158045203-d93180c2-744c-4e17-af9b-d90b9fa8df72.png)
> 
> I am not sure if this has been patched already. 

fish shell will parse the yaml into slices，and regard it as completions，it is not a complete one，fish shell regard Enter as completion delimiter

like this
![image](https://user-images.githubusercontent.com/12210746/145072583-9a420b90-8dd9-4541-b679-51ac48045d6a.png)

**Comment by DragonflyRobotics** [comment]
@NextAlone, I am not sure I completely understood what you just said. Do you mean that when we try to autocomplete, we should see the parsed yaml with the \ instead of what you see in my image?

**Comment by NextAlone** [comment]
@DragonflyRobotics yes, it maybe not support multi line completion, and,  if there any possibility to convert it to json?

**Comment by NextAlone** [comment]
and in https://github.com/fish-shell/fish-shell/issues/5433, shows that it dont support wrap completion with quote but backsplash

**Comment by sloretz** [comment]
ROS 1 is end-of-life (EOL) as of today, May 31st 2025. I am archiving this repository because:

* it only supports ROS 1
* it isn't needed anymore in ROS 2

[If you still rely on ROS 1, read this page to learn about your options.](https://bit.ly/NoeticEOL)
